Issue 2369: Update controller table queries and operations to handle variable sized rows

shiveshr opened the pull request:
**Change log description**
1. Add index to segment and history table and update all queries such that indexes are used for optimal access. 
2. New serialization for SegmentRecord and HistoryRecord. 

**Purpose of the change**
Fixes #2369

**What the code does**
Adds new Segment Index and Replaces index table with a history index. 
The index allows for variable sized rows in the table. 
All queries rely on indexed entries in the code. We always update Index before the entry in the table. Because repopulating the entry from the tabled record could be error prone. 
This becomes more interesting when we add multiple rows into segment table. If in a previous iteration only index was updated while the table entry was not updated, we erase all stray index entries and overwrite them with new entries that match the record that we write in the actual table. 

**How to verify it**
New unit tests added.

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java:
```diff
@@ -498,10 +501,45 @@ public void refresh() {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                })
+                .thenApply(x -> cache.invalidateCache(segmentIndexPath));
```
thenRun

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/store/stream/tables/SegmentRecord.java:
```diff
@@ -9,62 +9,87 @@
+        return Optional.empty();
     }
 
+    @SneakyThrows
```
`SneakyThrows(IOException.class)`

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/store/stream/tables/HistoryRecord.java:
```diff
@@ -9,238 +9,173 @@
 
-    private static int tailLengthOffset(int segmentCount) {
-        return scaleTimeOffset(segmentCount) + Long.BYTES;
+    @SneakyThrows
```
`SneakyThrows(IOException.class)`

And really, it can throw a legitimate IOException at you, so I don't understand why you want to suppress its declaration.

shiveshr replied:
Its a checked exception that we will have to handle everywhere else.

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/store/stream/tables/SegmentRecord.java:
```diff
@@ -9,62 +9,87 @@
-    private static double toDouble(byte[] b, int offset) {
-        return Double.longBitsToDouble(BitConverter.readLong(b, offset));
+    @SneakyThrows
+    byte[] toByteArray() {
```
In order to avoid all this byte array copying everywhere, I recommend changing this to return `ArrayView` and keep using it until you really have no choice but copy it to a byte array. For instance, many APIs allow taking a byte array with offset + length, so no need for any additional copies.

shiveshr replied:
done

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java:
```diff
@@ -71,111 +71,126 @@ public static Segment getSegment(final int number, final byte[] segmentTable) {
+
+                segments.addAll(fromStreamCut);
+                // put remaining segments as those that dont overlap with ones taken from streamCut.
+                segments.addAll(record.getSegments().stream().filter(x -> fromStreamCut.stream().noneMatch(y ->
```
1. No need for a collector here. Simply invoke `.forEach(segments::add)` which will prevent the additional list creation.

2. Also, you are looping several times over `fromStreamCut` and invoking `getSegment(y, segmentIndex, segmentTable)`. I suggest you pre-cache these values in a local map.

shiveshr replied:
> Also, you are looping several times over fromStreamCut and invoking getSegment(y, segmentIndex, segmentTable). I suggest you pre-cache these values in a local map.

getSegment is an O(1) fetch from the segment and index blobs. Its not a remote call.

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java:
```diff
@@ -308,23 +332,23 @@ public static long getSizeTillStreamCut(final byte[] indexTable, final byte[] hi
         if (!historyRecordOpt.isPresent()) {
-            // cant compute predecessors because the creation event is not present in history table yet.
+            // cant compute predecessors because the segment creation entry is not present in history table yet.
             return new ArrayList<>();
```
`Collections.emptyList()`

here and below

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java:
```diff
@@ -340,49 +364,65 @@ public static long getSizeTillStreamCut(final byte[] indexTable, final byte[] hi
-                                segmentStream.write(new SegmentRecord(startingSegmentNumber + newRange,
-                                        creationEpoch, timeStamp, newRanges.get(newRange).getKey(), newRanges.get(newRange).getValue())
+                                int offset = segmentStream.size();
+                                segmentStream.write(new SegmentRecord(startingSegmentNumber + x,
```
This is an excellent example of `toByteArray` being redundant. If you had returned `ArrayView`, you would have called `write(av.array(), av.arrayOffset(), av.length())` which would only copy the data once, vs allocating an array just so that you can copy it wholly.

shiveshr commented on controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java:
```diff
@@ -276,6 +278,55 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
+                result.completeExceptionally(StoreException.create(StoreException.Type.DATA_NOT_FOUND,
+                        "Segment index for stream: " + getName()));
+            } else if (segmentIndex.getVersion().equals(data.getVersion())) {
+                segmentIndex = new Data<>(Arrays.copyOf(data.getData(), data.getData().length), data.getVersion() + 1);
```
done

shiveshr commented:
@andreipaduroiu 

>    I thought we split this PR into two, when only the second one contained serialization changes. I see quite a bit of Versioned Serialization in here.

The two classes that we were changing (SegmentRecord and HistoryRecord) we updated their serialization here as well. For all other objects (around 10 of the remaining ones) we have a separate follow up PR to address it. 

>    I strongly recommend changing the toByteArray to something like toArrayView. There are countless cases where a new byte array is unnecessarily created & copied from somewhere just so that it is again copied somewhere else. ArrayView would prevent many of those copies from happening.

Done

shiveshr commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Use custom serialization with version on controller metadata](https://trello.com/c/rmd8xTsI/15-use-custom-serialization-with-version-on-controller-metadata)

shiveshr commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Use custom serialization with version on controller metadata](https://trello.com/c/rmd8xTsI/15-use-custom-serialization-with-version-on-controller-metadata)

shiveshr commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Use custom serialization with version on controller metadata](https://trello.com/c/rmd8xTsI/15-use-custom-serialization-with-version-on-controller-metadata)

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java:
```diff
@@ -71,111 +72,126 @@ public static Segment getSegment(final int number, final byte[] segmentTable) {
+    /**
+     * Method to validate a given stream Cut.
+     * A stream cut is valid if it covers the entire key space without any overlaps in ranges for segments that form the
+     * streamcut. It throws Invalid argument exception if the supplied stream cut does not satisfy the invariants.
```
Add a @link to the exception reference.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/tables/serializers/HistoryRecordSerializer.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Please do not reformat the license header.

shiveshr replied:
Done

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/tables/serializers/HistoryRecordSerializer.java:
```diff
@@ -0,0 +1,49 @@
+import java.io.IOException;
+import java.util.ArrayList;
+
+public class HistoryRecordSerializer extends VersionedSerializer.WithBuilder<HistoryRecord, HistoryRecord.HistoryRecordBuilder> {
```
Where are you unit-testing the seriaization/deserialization? I think this is being exercised implicitly, and I'd rather see some explicit unit tests that focus on it alone, please.

shiveshr replied:
The unit test i had added for all serialization/deserialization will be included in the follow up PR.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java:
```diff
@@ -276,6 +278,56 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
+
+            return CompletableFuture.completedFuture(copy(this.segmentIndex));
+        }
+
```
Eliminate unnecessary new line.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java:
```diff
@@ -289,8 +341,9 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
-        return getSegmentTable()
-                .thenApply(x -> TableHelper.getSegment(number, x.getData()));
+        return getSegmentIndex()
+            .thenCompose(segmentIndex -> getSegmentTable()
```
Indentation is incorrect.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java:
```diff
@@ -331,45 +385,47 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
         Preconditions.checkNotNull(updated.getData());
 
         final CompletableFuture<Void> result = new CompletableFuture<>();
+        Data<Integer> copy = updatedCopy(updated);
```
Currently, this copy is performed under a lock, but this PR is proposing to copy outside the lock. Could you justify why this change is correct?

shiveshr replied:
We are copying the input that we receive because it is mutable and outside the control of this class. 
We dont need to copy it within a lock though as what we want to protect under lock is when we flip the reference to this new copied object.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java:
```diff
@@ -610,6 +666,7 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
     @Override
     CompletableFuture<Void> updateMarkerData(int segmentNumber, Data<Integer> data) {
         CompletableFuture<Void> result = new CompletableFuture<>();
+        Data<Integer> next = updatedCopy(data);
```
Same here about copying under a lock.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java:
```diff
@@ -170,11 +171,12 @@ public String getScopeName() {
-                                .thenApply(index -> TableHelper.computeTruncationRecord(index.getData(), history.getData(),
-                                        segment.getData(), streamCut, truncationRecord))));
+        return getHistoryIndexFromStore()
+                .thenCompose(historyIndex -> getHistoryTableFromStore()
```
I think it is better to align the beginning of the lines rather than cascading like this.

shiveshr replied:
This is because thenCompose is being called on the method call inside thenCompose block. 
This is done so that the value from each future is available at all those below it.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java:
```diff
@@ -622,27 +640,33 @@ private long findSegmentSplitsMerges(List<Segment> referenceSegmentsList, List<S
 
-        final byte[] updated = TableHelper.updateSegmentTable(nextSegmentNumber, newEpoch,
-                segmentTable.getData(), newRanges, segmentCreationTimestamp);
+        // Note: if segment index was updated in an earlier attempt but segment was not, we need to overwrite previous index update!
```
`... overwrite the previous index update` it is missing `the`.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java:
```diff
@@ -622,27 +640,33 @@ private long findSegmentSplitsMerges(List<Segment> referenceSegmentsList, List<S
-                segmentTable.getData(), newRanges, segmentCreationTimestamp);
+        // Note: if segment index was updated in an earlier attempt but segment was not, we need to overwrite previous index update!
+        // this is because new offsets may be different. we cannot update segment table before index because then
+        // we would not be able to read from segment table as we dont know the starting offsets for segments.
```
`dont` -> `don't`

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java:
```diff
@@ -695,35 +719,37 @@ private long findSegmentSplitsMerges(List<Segment> referenceSegmentsList, List<S
+                                            if (e != null) {
+                                                Throwable ex = Exceptions.unwrap(e);
+                                                if (ex instanceof StoreException.DataNotEmptyException) {
+                                                    // cant delete as there are transactions still running under epoch node
```
`cant` -> `can't`

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java:
```diff
@@ -1138,17 +1168,18 @@ private void checkState(State currState, State expectedState) {
                             if (lastRecord.getEpoch() > activeEpoch) {
-                                boolean idempotent = lastRecord.isPartial() &&
-                                        lastRecord.getSegments().containsAll(createdSegments) && lastRecord.getEpoch() == newEpoch;
+                                boolean idempotent = lastRecord.isPartial() && lastRecord.getSegments().containsAll(createdSegments);
```
This term has been removed `lastRecord.getEpoch() == newEpoch`, why?

shiveshr replied:
it is already captured in the if condition `lastRecord.getEpoch() > activeEpoch`. 
Let me try to recollect if there was any other specific reason for the same though.

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/tables/HistoryIndexRecord.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Class corresponding to a record/row in Index table.
+ * Each row is fixed size
+ * Row: [eventTime, pointer-into-history-table]
```
`eventTime` here is confusing, what do you mean with event time?

shiveshr replied:
EpochCreationTime or scale time..  Will update the comment

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java:
```diff
@@ -71,111 +72,126 @@ public static Segment getSegment(final int number, final byte[] segmentTable) {
-                            .collect(Collectors.toList()));
-                }
+                // case 3: overlap between requested epoch and stream cut.
+                // take segments from stream cut that are from or aftergit re this epoch.
```
`... from or after this epoch.`

fpj commented on controller/src/main/java/io/pravega/controller/store/stream/tables/HistoryIndexRecord.java:
```diff
@@ -0,0 +1,52 @@
+ * Each row is fixed size
+ * Row: [eventTime, pointer-into-history-table]
+ */
+public class HistoryIndexRecord {
```
Are we supposed to be getting rid of `IndexRecord` in this PR? I don't see it being removed.

shiveshr replied:
Index Record is replaced with HistoryIndexRecord
